src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorManager.java

```diff
@@ -83,14 +89,56 @@ public void vendorRepos(Path externalRepoRoot, ImmutableList<RepositoryName> rep
       // 3. Move the external repo to vendor dir. It's fine if this step fails or is interrupted,
       // because the marker file under external is gone anyway.
       FileSystemUtils.moveTreesBelow(repoUnderExternal, repoUnderVendor);
-      // 4. Rename to temporary marker file after the move is done.
+      // 4. Re-plant symlinks pointing a path under the external root to a relative path
+      // to make sure the vendor src keep working after being moved or output base changed
+      replantSymlinks(repoUnderVendor, externalRepoRoot);
+      // 5. Rename to temporary marker file after the move is done.
       tMarker.renameTo(markerUnderVendor);
-      // 5. Leave a symlink in external dir.
+      // 6. Leave a symlink in external dir.
       repoUnderExternal.deleteTree();
       FileSystemUtils.ensureSymbolicLink(repoUnderExternal, repoUnderVendor);
     }
   }
 
+  /**
+   * Replants the symlinks under the specified repository directory.
+   *
+   * Re-write symlinks that originally pointing to a path under the external root to a relative path pointing
+   * to an external root symlink under the vendor directory.
+   *
+   * @param repoUnderVendor The path to the repository directory under the vendor directory.
+   * @param externalRepoRoot The path to the root of external repositories.
+   * @throws IOException If an I/O error occurs while replanting the symlinks.
+   */
+  private void replantSymlinks(Path repoUnderVendor, Path externalRepoRoot) throws IOException {
+    try {
+      Collection<Path> symlinks = FileSystemUtils.traverseTree(repoUnderVendor, Path::isSymbolicLink);
+      Path externalSymlinkUnderVendor = vendorDirectory.getChild(EXTERNAL_ROOT_SYMLINK_NAME);
+      FileSystemUtils.ensureSymbolicLink(externalSymlinkUnderVendor, externalRepoRoot);
+      for (Path symlink : symlinks) {
+        PathFragment target = symlink.readSymbolicLink();
+        if (!target.startsWith(externalRepoRoot.asFragment())) {
+          // TODO: print a warning for absolute symlinks?
+          continue;
+        }
+        PathFragment newTarget = PathFragment
+            .create("../".repeat(symlink.relativeTo(vendorDirectory).segmentCount() - 1))
+            .getRelative(EXTERNAL_ROOT_SYMLINK_NAME)
+            .getRelative(target.relativeTo(externalRepoRoot.asFragment()));
+        if (OS.getCurrent() == OS.WINDOWS){
+          // On Windows, FileSystemUtils.ensureSymbolicLink always resolves paths to absolute path.
+          // Use Files.createSymbolicLink here instead to preserve relative target path.
+          symlink.delete();
+          Files.createSymbolicLink(java.nio.file.Path.of(symlink.getPathString()), java.nio.file.Path.of(newTarget.getPathString()));
```

Comment:
Are we also rewriting junctions to symlinks here? I'm asking because the former wouldn't require admin rights, but the latter may.

Reply:
Nope, the `Files.` class doesn't offer any way to create Junction via API. But if we can solve the `FileSystemUtils.ensureSymbolicLink` problem, it should be able to do that.

Reply:
Since junction only supports absolute path, this has to be a symlink. I updated the doc to mention this.